server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java

```diff
@@ -27,8 +27,8 @@ public ClusterInfoSimulator(ClusterInfo clusterInfo) {
         this.leastAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeLeastAvailableDiskUsages());
         this.mostAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeMostAvailableDiskUsages());
         this.shardSizes = new HashMap<>(clusterInfo.shardSizes);
-        this.shardDataSetSizes = new HashMap<>(clusterInfo.shardDataSetSizes);
-        this.dataPath = new HashMap<>(clusterInfo.dataPath);
+        this.shardDataSetSizes = clusterInfo.shardDataSetSizes;
+        this.dataPath = clusterInfo.dataPath;
```

Comment:
Those 2 are immutable maps at this point. They would be reused by Map.copyOf. This should help a lot as shardDataSetSizes size is proportional to a shard count.

Reply:
I think I'd prefer to use `Map.copyOf()` here just to be explicit that we're not modifying anything in `clusterInfo`. (we could reasonably `assert this.shardDataSetSizes == clusterInfo.shardDataSetSizes` to protect against a performance regression)